README.md
# <img align="right" src="http://vlad-ignatov.github.io/react-numeric-input/examples/v2.0.0/screenshot.png" width="123"/>React Numeric Input

[![Build Status](https://travis-ci.org/vlad-ignatov/react-numeric-input.svg?branch=master)](https://travis-ci.org/vlad-ignatov/react-numeric-input)

Number input component that can replace the native number input which is not yet
very well supported and where it is, it does not have the same appearance across
the browsers. Additionally this component offers more flexible options and can
be used for any values (differently formatted representations of the internal
numeric value).

![Demo](https://i.imgur.com/uUnxO73.gif)
### [**Live demo**](http://vlad-ignatov.github.io/react-numeric-input/)

## Installation
```sh
npm install react-numeric-input --save
```
Then in your scripts:
```js
// es6
import NumericInput from 'react-numeric-input';
// or es5
var NumericInput = require('react-numeric-input');
// or TypeScript
import * as NumericInput from "react-numeric-input";
```

## Usage
##### Minimal Usage:
This will behave exactly like `<input type="number">`. It will create an empty
numeric input that starts changing from zero. The difference that this works on
any browser and does have the same appearance on each browser.
```jsx
<NumericInput/>
// or:
<NumericInput className="form-control"/>
```

##### Typical Usage
Most of the time you will need to specify `min`, `max` and `value`:
```jsx
<NumericInput min={0} max={100} value={50}/>
```

#### Working with floats
You can use `step` and `precision` props to make your input working with
floating point numbers:
```jsx
<NumericInput step={0.1} precision={2} value={50.3}/>
```

#### Snap to step
If you want your component to "snap" to the closest step value while incrementing
or decrementing (up/down buttons or arrow keys) you can use the `snap` prop:
```jsx
<NumericInput step={0.5} precision={2} value={50.3} snap/>
```

#### Strict vs Loose Mode
You can type any value in the input as long as it is in focus. On blur, or when
you attempt to increment/decrement it, the value will be converted to number.
If you don't want this behaviour, pass `strict` in the props and any value that
cannot be converted to number will be rejected immediately.

#### Custom format
By default the component displays the value number as is. However, you can
provide your own `format` function that will be called with the numeric value
and is expected to return the string that will be rendered in the input:
```jsx
function myFormat(num) {
    return num + '$';
}
<NumericInput precision={2} value={50.3} step={0.1} format={myFormat}/>
```
Please note that the example above is fine but in most situations if you have custom
`format` function you will also need to provide custom `parse` function that is able to
convert whatever the `format` returns back to numeric value. In the example above the
built-in `parse` function will strip the "$" suffix because internally it uses `parseFloat`.
However, if the `format` function was returning `"$" + num`, then the `parse` function
should do something like:
```js
function parse(stringValue) {
    return stringValue.replace(/^\$/, "");
}
```

## Props
Name              | Type                                | Default
------------------|-------------------------------------|:-------:
**value**         |`number` or `string`                 |`""` which converts to 0
**min**           |`number` or `function`               |`Number.MIN_SAFE_INTEGER`
**max**           |`number` or `function`               |`Number.MAX_SAFE_INTEGER`
**step**          |`number` or `function`               | 1
**precision**     |`number` or `function`               | 0
**parse**         |`function`                           | parseFloat
**format**        |`function`                           | none
**className**     |`string`                             | none
**disabled**      |`boolean`                            | none
**readOnly**      |`boolean`                            | none
**style**         |`object` or `false`                  | none
**size**          |`number` or `string`                 | none
**mobile**        |`true`, `false`, 'auto' or `function`|`auto`
**snap**          |`boolean`                            | none (false)
**componentClass**|`string` or `function`               |`"input"`
**strict**        |`boolean`                            |`false`

Any other option is passed directly the input created by the component. Just
don't forget to camelCase the attributes. For example `readonly` must be `readOnly`.
See examples/index.html for examples.

## Event Callbacks
You can pass callback props like `onClick`, `onMouseOver` etc. and they will be
attached to the input element and React will call them with `null` scope and the corresponding event. However, there are few special cases to be aware of:

* `onChange`  - Called with `valueAsNumber`, `valueAsString` and the `input` element. The `valueAsNumber` represents the internal numeric value while `valueAsString` is the same as the input value and might be completely different from the numeric one if custom formatting is used.
* `onInvalid` - Will be called with `errorMessage`, `valueAsNumber` and `valueAsString`.
* `onValid`   - There is no corresponding event in browsers. It will be called when the component transitions from invalid to valid state with the same arguments as onChange: `valueAsNumber` and `valueAsString`.

## Styling
The component uses inline styles which you can customize. The `style` prop is not added
directly to the component but instead it is a container for styles which you can overwrite.
For example
```xml
<NumericInput style={{
	input: {
		color: 'red'
	}
}}>
```
You can modify the styles for everything including states like `:hover`, `:active` and
`:disabled`. Take a look at the source to see what styles are supported. Also, the style is
stored as static class property so that you can change it and affect all the components
from your script. Example:
```js
import NumericInput from 'react-numeric-input';
NumericInput.style.input.color = 'red';
```

Finally, you can still use CSS if you want. Each component's root element has the
`react-numeric-input` class so that it is easy to find these widgets on the page. However,
keep in mind that because of the inline styles you might need to use `!important` for some
rules unless you pass `style={false}` which will disable the inline styles and you will
have to provide your own CSS styles for everything. Example:
```css
.react-numeric-input input {
	color: red;
}
```

## Keyboard navigation
* You can use <kbd>⬆</kbd> and <kbd>⬇</kbd> arrow keys to increment/decrement the input value.
* <kbd>Ctrl + ⬆</kbd> or <kbd>⌘ + ⬆</kbd> and <kbd>Ctrl + ⬇</kbd> or <kbd>⌘ + ⬇</kbd> to use smaller step (`step / 10`).
  Note that this will only work if you have specified a `precision` option that supports it.
* <kbd>Shift + ⬆</kbd> and <kbd>Shift + ⬇</kbd> to use bigger step (`step * 10`).

## Integration with external scripts
This component aims to provide good integration not only with React but with any third party script
that might want to work with it on the current page.

### getValueAsNumber()
The native number inputs have special property called `valueAsNumber`. It provides access to the
value as number to be used by scripts. In this react component this becomes even more desirable as
the display value might be formatted and have nothing in common with the underlying value meaning
that one might need to call parse to find out what the numeric value is. For that reason this
component exposes `getValueAsNumber()` method on the input element. Also keep in mind
that this really returns a number (float) so it might be different from the displayed value. For
example an input showing "12.30" will have `getValueAsNumber()` returning `12.3` and if the input
is empty the result would be `0`.

### setValue()
An external script that does not "understand" React can still work with this
component by reading the `getValueAsNumber()` or by calling the `setValue()`
method exposed on the input element. Here is an example with jQuery:
```js
$('input[name="some-input"]')[0].setValue('123mph');
```
Calling this method will invoke the component's `parse` method with the provided
argument and then it will `setState` causing the usual re-rendering.

## function props
In rare cases it might be better to "decide" what the value of certain prop is at
runtime without having to through the entire ceremony of redux, flux, or whatever
will make your component to be re-rendered with other props. For example one might
want to have a variable `step` prop based on the current input value and the change
direction:
```js
<NumericInput step={(component, direction) => {
	// for values smaller than 10 the step is 0.1
	// for values greater than 10 the step is 0.01
	return component.state.value < 10 ? 0.1 : 0.01

	// or have different step depending on the direction
	return direction === NumericInput.DIRECTION_UP ? 0.01 : 0.1;

	// or just obtain it from somewhere
	return window.outerWidth % 100 + 1;
}}>
```
The props that support being a function are currently `min`, `max`, `step` and
`precision`. All those function will be passed the component instance as argument
and the `step` will also receive the direction as second parameter.


## License
MIT


CHANGELOG.md
2.2.3
--------------------------------------------------------------------------------
* Small fix for a bug introduced in 2.2.2.

2.2.2
--------------------------------------------------------------------------------
* style={false} produces eslint warning so it is deprecated now. There is new `noStyle` boolean prop for that.
* Fixes the case where string value reappears in input when mouse hovers over the up/down button after clicking away.
* Document the `strict` prop.
* A readOnly input is almost like disabled one now. The only difference is that you can copy the value.
* Various small code improvements.

2.2.1
--------------------------------------------------------------------------------
* The input will now use "fontSize: inherit" style.
* Added the componentClass prop (thanks to @Yukioru)
* Fixed small issues in tests
* Added live demo preview (thanks to @ipiranhaa)
* Accept React 16 as peer dependency

2.2.0
--------------------------------------------------------------------------------
* If the component gets re-rendered with new value the onChange callback will not be called.
* If the component is rendered with invalid string value prop the input will render empty instead of having value of `0`.
* Fixed an issue where preventDefault in "onTouchStart" causes double step to happen.
* Added the `strict` prop. In strict mode the value is converted to number (and formated) while the user types. In loose mode (the default mode) you can type anything and the value gets reformatted on blur.
* Added e2e tests.
* The precision prop can also be a function that can decide what precision to use at runtime.
* The min, max and step props can also be functions.

2.1.0
--------------------------------------------------------------------------------
* Added `snap` prop to make the value snap to the closest step when using buttons or up/down keys to change it.
* Pass a reference to the input element to the onChange callbacks
* Fixed an issue preventing the input from being editable when it's entire value is selected
* Fixed an issue that might cause the component to loose its value if re-rendered without value prop

2.0.9
--------------------------------------------------------------------------------
* Field now accepts `"-."`, `"+.`", `"."`, `"-"`, and `"+"` at beginning of input value.
* Merge some pull requests
* Fixed backspace on float values
* Use external prop-types package
* Use react and react-dom v15.6.1

2.0.8
--------------------------------------------------------------------------------
* Fixed `React.PropTypes` reference warnings

2.0.7
--------------------------------------------------------------------------------
* Fixed the focus/blur issues. Thanks to https://github.com/vlad-ignatov/react-numeric-input/pull/25
* onChange gets called when input content is deleted #27

2.0.6
--------------------------------------------------------------------------------
* Use peerDependencies to specify the react versions


2.0.5
--------------------------------------------------------------------------------
* Compatible with React 15.2 and 15.3
* Preserve the selection between re-rendering
* Fix a bug in handling the step prop
* Loose formatting while the user is typing and strict on blur
* Various other fixes

2.0.4
--------------------------------------------------------------------------------
* Make it compatible with React 15
* Inline styles can be disabled by passing style={false}
* Various small fixes and improvements


2.0.3
--------------------------------------------------------------------------------
* Provide more useful data to event callbacks like onChange, onInvalid etc.
* Make sure that onBlur, onFocus and onKeyDown are called the same way React would
call them.


2.0.2
--------------------------------------------------------------------------------
* Add support for callback props onChange, onInvalid, onValid and more...
* defaultValue can also be used instead of value
* Make use of HTML5 input validation if available


2.0.1
--------------------------------------------------------------------------------
* Move react and react-dom to devDependencies so that npm does not install them


2.0.0
--------------------------------------------------------------------------------
* Switch to inline styles
* Better integration with external scripts
* Better integration with build tools
* Support for touch devices


1.0.0
--------------------------------------------------------------------------------
* Initial release


examples/examples.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory(require("React"), require("ReactDOM"));
	else if(typeof define === 'function' && define.amd)
		define(["React", "ReactDOM"], factory);
	else {
		var a = typeof exports === 'object' ? factory(require("React"), require("ReactDOM")) : factory(root["React"], root["ReactDOM"]);
		for(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];
	}
})(this, function(__WEBPACK_EXTERNAL_MODULE_1__, __WEBPACK_EXTERNAL_MODULE_2__) {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};

/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {

/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;

/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			exports: {},
/******/ 			id: moduleId,
/******/ 			loaded: false
/******/ 		};

/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

/******/ 		// Flag the module as loaded
/******/ 		module.loaded = true;

/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}


/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;

/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;

/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";

/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

	"use strict";

	var _react = __webpack_require__(1);

	var _react2 = _interopRequireDefault(_react);

	var _reactDom = __webpack_require__(2);

	var _reactDom2 = _interopRequireDefault(_reactDom);

	var _Demo = __webpack_require__(3);

	var _Demo2 = _interopRequireDefault(_Demo);

	function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

	// import NumericInput from '../index.js';


	$(function () {
	    $('script.jsx').each(function (i, s) {
	        var div = $('<div/>'),
	            props = Function('return (' + $(s).text() + ')')();
	        $(s).replaceWith(div);
	        var widget = _reactDom2.default.render(_react2.default.createElement(NumericInput, props), div[0]);
	        div.data("widget", widget);
	    });

	    _reactDom2.default.render(_react2.default.createElement(_Demo2.default, null), $('.demo')[0]);

	    hljs.configure({ useBR: false });

	    $('.code').each(function (i, block) {
	        hljs.highlightBlock(block);
	    });
	}); /* global $, hljs, NumericInput */

/***/ }),
/* 1 */
/***/ (function(module, exports) {

	module.exports = __WEBPACK_EXTERNAL_MODULE_1__;

/***/ }),
/* 2 */
/***/ (function(module, exports) {

	module.exports = __WEBPACK_EXTERNAL_MODULE_2__;

/***/ }),
/* 3 */
/***/ (function(module, exports) {

	"use strict";

	Object.defineProperty(exports, "__esModule", {
	    value: true
	});

	var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

	var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

	function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

	function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

	function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

	function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

	/* global $, hljs, NumericInput, React */
	var Demo = function (_React$Component) {
	    _inherits(Demo, _React$Component);

	    function Demo() {
	        var _ref;

	        _classCallCheck(this, Demo);

	        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
	            args[_key] = arguments[_key];
	        }

	        // var that = this;
	        var _this = _possibleConstructorReturn(this, (_ref = Demo.__proto__ || Object.getPrototypeOf(Demo)).call.apply(_ref, [this].concat(args)));

	        _this.state = {
	            inputProps: {
	                name: { value: "whatever", on: false },
	                className: { value: "form-control", on: true },
	                value: { value: 50, on: true },
	                min: { value: 0, on: true },
	                max: { value: 100, on: true },
	                step: { value: 1, on: true },
	                precision: { value: 0, on: true },
	                size: { value: 5, on: true },
	                maxLength: { value: 2, on: false },
	                disabled: { value: true, on: false },
	                readOnly: { value: true, on: false },
	                mobile: { value: true, on: false },
	                required: { value: true, on: false },
	                noValidate: { value: true, on: false },
	                pattern: { value: "[0-9].[0-9][0-9]", on: false },
	                title: { value: "The title attr", on: false },
	                snap: { value: true, on: false },
	                inputmode: { value: "numeric", on: false
	                    // library
	                } }
	        };
	        return _this;
	    }

	    _createClass(Demo, [{
	        key: "componentDidUpdate",
	        value: function componentDidUpdate() {
	            hljs.highlightBlock(this.refs.code);
	        }
	    }, {
	        key: "toggleProp",
	        value: function toggleProp(propName) {
	            this.state.inputProps[propName].on = !this.state.inputProps[propName].on;
	            this.setState(this.state);
	        }
	    }, {
	        key: "setProp",
	        value: function setProp(propName, event) {
	            var val = event.target ? event.target.value : event;
	            this.state.inputProps[propName].value = val;
	            this.setState(this.state);
	        }
	    }, {
	        key: "onChange",
	        value: function onChange(x) {
	            this.state.inputProps.value.value = x === null ? "" : x;
	            if (this.state.inputProps.value.on) {
	                this.setState(this.state);
	            }
	        }
	    }, {
	        key: "onInvalid",
	        value: function onInvalid(message) {
	            // console.log("Invalid", message)
	            $(this.refs.errorMessage).text(message || "Unknown error");
	        }
	    }, {
	        key: "onValid",
	        value: function onValid() {
	            // console.log("Valid")
	            $(this.refs.errorMessage).empty();
	        }
	    }, {
	        key: "renderCode",
	        value: function renderCode() {
	            var out = '<NumericInput ';
	            var hasProps = false;

	            for (var propName in this.state.inputProps) {
	                if (this.state.inputProps[propName].on && !this.state.inputProps[propName].hidden) {
	                    var val = this.state.inputProps[propName].value;
	                    out += "\n\t" + propName;
	                    if (val !== true) {
	                        out += '=' + (typeof val == 'string' ? "\"" + val + "\" " : "{ " + val + " } ");
	                    }
	                    hasProps = true;
	                }
	            }

	            if (hasProps) {
	                out += '\n';
	            }

	            out += '/>';

	            return React.createElement(
	                "div",
	                { className: "code js", ref: "code", style: { minHeight: 379 } },
	                out
	            );
	        }
	    }, {
	        key: "renderPropEditors",
	        value: function renderPropEditors(config) {
	            var _this2 = this;

	            return config.map(function (props, propName) {
	                var editor = null;

	                var type = props.type,
	                    name = props.name,
	                    rest = _objectWithoutProperties(props, ["type", "name"]);

	                if (type == 'text') {
	                    editor = React.createElement("input", {
	                        type: "text",
	                        className: "form-control input-sm",
	                        value: _this2.state.inputProps[name].value,
	                        onChange: _this2.setProp.bind(_this2, name)
	                    });
	                } else if (type == "number") {
	                    editor = React.createElement(NumericInput, _extends({
	                        className: "form-control input-sm",
	                        value: _this2.state.inputProps[name].value,
	                        onChange: _this2.setProp.bind(_this2, name)
	                    }, rest));
	                }

	                return React.createElement(
	                    "tr",
	                    { key: propName },
	                    React.createElement(
	                        "td",
	                        { className: "unselectable" },
	                        React.createElement(
	                            "label",
	                            { style: { display: "block" } },
	                            React.createElement("input", {
	                                type: "checkbox",
	                                checked: _this2.state.inputProps[name].on,
	                                onChange: _this2.toggleProp.bind(_this2, name)
	                            }),
	                            "\xA0",
	                            name
	                        )
	                    ),
	                    React.createElement(
	                        "td",
	                        null,
	                        editor
	                    )
	                );
	            });
	        }
	    }, {
	        key: "render",
	        value: function render() {
	            var inputProps = {};
	            for (var propName in this.state.inputProps) {
	                if (this.state.inputProps[propName].on) {
	                    inputProps[propName] = this.state.inputProps[propName].value;
	                }
	                // else {
	                //     inputProps[propName] = null
	                // }
	            }

	            return React.createElement(
	                "div",
	                { className: "row" },
	                React.createElement(
	                    "div",
	                    { className: "col-xs-6" },
	                    React.createElement(
	                        "div",
	                        { className: "panel panel-default" },
	                        React.createElement(
	                            "div",
	                            { className: "panel-heading" },
	                            "Props"
	                        ),
	                        React.createElement(
	                            "table",
	                            { className: "table table-striped table-condensed" },
	                            React.createElement(
	                                "colgroup",
	                                null,
	                                React.createElement("col", { width: 169 }),
	                                React.createElement("col", null)
	                            ),
	                            React.createElement(
	                                "thead",
	                                null,
	                                React.createElement(
	                                    "tr",
	                                    null,
	                                    React.createElement(
	                                        "th",
	                                        null,
	                                        "name"
	                                    ),
	                                    React.createElement(
	                                        "th",
	                                        null,
	                                        "value"
	                                    )
	                                )
	                            )
	                        ),
	                        React.createElement(
	                            "div",
	                            { style: {
	                                    overflow: 'auto',
	                                    maxHeight: 452
	                                } },
	                            React.createElement(
	                                "table",
	                                { className: "table table-striped table-condensed" },
	                                React.createElement(
	                                    "colgroup",
	                                    null,
	                                    React.createElement("col", { width: 169 }),
	                                    React.createElement("col", null)
	                                ),
	                                React.createElement(
	                                    "tbody",
	                                    null,
	                                    this.renderPropEditors([{ name: "name", type: "text" }, { name: "className", type: "text" }, { name: "value", type: "text" }, { name: "min", type: "number" }, { name: "max", type: "number" }, { name: "step", type: "number", min: 0.001, step: 0.1, precision: 3 }, { name: "precision", type: "number", min: 0, max: 20 }, { name: "size", type: "number", min: 0, max: 60 }, { name: "maxLength", type: "number", min: 0, max: 20 }, { name: "disabled", type: "bool" }, { name: "readOnly", type: "bool" }, { name: "mobile", type: "bool" }, { name: "pattern", type: "text" }, { name: "title", type: "text" }, { name: "required", type: "bool" }, { name: "noValidate", type: "bool" }, { name: "inputmode", type: "text" }, { name: "snap", type: "bool" }])
	                                )
	                            )
	                        )
	                    )
	                ),
	                React.createElement(
	                    "div",
	                    { className: "col-xs-6" },
	                    React.createElement(
	                        "div",
	                        { className: "panel panel-primary" },
	                        React.createElement(
	                            "div",
	                            { className: "panel-heading" },
	                            "Preview"
	                        ),
	                        React.createElement(
	                            "div",
	                            { className: "panel-body" },
	                            React.createElement(
	                                "div",
	                                { ref: "example" },
	                                React.createElement(NumericInput, _extends({}, inputProps, {
	                                    onChange: this.onChange.bind(this),
	                                    onInvalid: this.onInvalid.bind(this),
	                                    onValid: this.onValid.bind(this),
	                                    value: inputProps.value === null ? undefined : inputProps.value || ""
	                                })),
	                                React.createElement(
	                                    "div",
	                                    { className: "help-block" },
	                                    React.createElement("span", { ref: "errorMessage", className: "text-danger" })
	                                )
	                            ),
	                            React.createElement("hr", null),
	                            this.renderCode()
	                        )
	                    )
	                )
	            );
	        }
	    }]);

	    return Demo;
	}(React.Component);

	exports.default = Demo;

/***/ })
/******/ ])
});
;